cmd/uplink/cmd/cp.go

```diff
@@ -4,107 +4,209 @@
+		return err
+	}
+
+	fmt.Printf("Created: %s\n", destObj.Path)
+
+	return nil
 }
 
-func copy(cmd *cobra.Command, args []string) (err error) {
-	ctx := process.Ctx(cmd)
+// download downloads s3 compatible object args[0] to args[1] on local machine
+func download(ctx context.Context, bs buckets.Store, srcObj *url.URL, destFile string) error {
+	if srcObj.Scheme == "" {
+		return fmt.Errorf("Invalid source")
+	}
 
-	if len(args) == 0 {
-		return errs.New("No file specified for copy")
+	o, err := bs.GetObjectStore(ctx, srcObj.Host)
+	if err != nil {
+		return err
 	}
 
-	if len(args) == 1 {
-		return errs.New("No destination specified")
+	if fi, err := os.Stat(destFile); err == nil && fi.IsDir() {
+		destFile = filepath.Join(destFile, filepath.Base(srcObj.Path))
 	}
 
-	so, err := getStorjObjects(ctx, cpCfg)
+	f, err := os.Create(destFile)
 	if err != nil {
 		return err
 	}
+	defer utils.LogClose(f)
 
-	u, err := url.Parse(args[0])
+	rr, _, err := o.Get(ctx, paths.New(srcObj.Path))
 	if err != nil {
 		return err
 	}
+	defer utils.LogClose(rr)
 
-	if u.Scheme == "" {
-		f, err := os.Open(args[0])
-		if err != nil {
-			return err
-		}
+	r, err := rr.Range(ctx, 0, rr.Size())
+	if err != nil {
+		return err
+	}
+	defer utils.LogClose(r)
 
-		fi, err := f.Stat()
-		if err != nil {
-			return err
-		}
+	_, err = io.Copy(f, r)
+	if err != nil {
+		return err
+	}
 
-		fr, err := hash.NewReader(f, fi.Size(), "", "")
-		if err != nil {
-			return err
-		}
+	fmt.Printf("Downloaded %s to %s\n", srcObj.Path, destFile)
 
-		defer func() { _ = f.Close() }()
+	return nil
+}
 
-		u, err = url.Parse(args[1])
-		if err != nil {
-			return err
-		}
+// copy copies s3 compatible object args[0] to s3 compatible object args[1]
+func copy(ctx context.Context, bs buckets.Store, srcObj *url.URL, destObj *url.URL) error {
+	o, err := bs.GetObjectStore(ctx, srcObj.Host)
+	if err != nil {
+		return err
+	}
 
-		oi, err := so.PutObject(ctx, u.Host, u.Path, fr, nil)
+	rr, _, err := o.Get(ctx, paths.New(srcObj.Path))
+	if err != nil {
+		return err
+	}
+	defer utils.LogClose(rr)
+
+	r, err := rr.Range(ctx, 0, rr.Size())
+	if err != nil {
+		return err
+	}
+	defer utils.LogClose(r)
+
+	if destObj.Host != srcObj.Host {
+		o, err = bs.GetObjectStore(ctx, destObj.Host)
 		if err != nil {
 			return err
 		}
+	}
 
-		fmt.Println("Bucket:", oi.Bucket)
-		fmt.Println("Object:", oi.Name)
+	meta := objects.SerializableMeta{}
+	expTime := time.Time{}
 
-		return nil
+	destObj.Path = cleanAbsPath(destObj.Path)
+	// if destination object name not specified, default to source object name
+	if strings.HasSuffix(destObj.Path, "/") {
+		destObj.Path = filepath.Join(destObj.Path, filepath.Base(srcObj.Path))
 	}
 
-	oi, err := so.GetObjectInfo(ctx, u.Host, u.Path)
+	_, err = o.Put(ctx, paths.New(destObj.Path), r, meta, expTime)
 	if err != nil {
 		return err
 	}
 
-	f, err := os.Create(args[1])
+	fmt.Printf("%s copied to %s\n", srcObj.Host+srcObj.Path, destObj.Host+destObj.Path)
```

Comment:
It would be better if we just print the complete urls instead of concatenating hosts and paths.

Reply:
Done